internal/orchestrator/orchestrator.go

```diff
@@ -89,56 +107,115 @@ func (o *Orchestrator) Run(ctx context.Context) {
+}
 
-	data, err := msg.Data()
+func (o *Orchestrator) pullEvents(ctx context.Context) {
+	// XXX: consider having a separate context for message retrieval
+	msgs, err := o.streamBroker.PullMsg(ctx, 1)
 	if err != nil {
 		o.logger.WithFields(
-			logrus.Fields{"err": err.Error(), "subject": msg.Subject()},
-		).Error("data unpack error")
+			logrus.Fields{"err": err.Error()},
+		).Debug("error fetching work")
+	}
+
+	for _, msg := range msgs {
+		if ctx.Err() != nil || o.concurrencyLimit() {
+			o.eventNak(msg)
+
+			return
+		}
+
+		// spawn msg process handler
+		o.syncWG.Add(1)
+
+		go func(msg events.Message) {
+			defer o.syncWG.Done()
+
+			atomic.AddInt32(&o.dispatched, 1)
+			defer atomic.AddInt32(&o.dispatched, -1)
+
+			o.processEvent(ctx, msg)
+		}(msg)
+	}
+}
+
+func (o *Orchestrator) processEvent(ctx context.Context, event events.Message) {
+	otelCtx, span := otel.Tracer(pkgName).Start(ctx, "processEvent")
+	defer span.End()
+
+	data, err := event.Data()
+	if err != nil {
+		o.ackEvent(event, errors.Wrap(ErrInvalidEvent, "data field error: "+err.Error()))
 
 		return
 	}
 
-	urn, err := msg.SubjectURN(data)
+	urn, err := event.SubjectURN(data)
 	if err != nil {
-		o.logger.WithFields(
-			logrus.Fields{"err": err.Error(), "subject": msg.Subject()},
-		).Error("error parsing subject URN in msg")
+		o.ackEvent(event, errors.Wrap(ErrInvalidEvent, "error parsing subject URN in msg: "+err.Error()))
 
 		return
 	}
 
 	if urn.ResourceType != ptypes.ServerResourceType {
-		o.logger.WithFields(
-			logrus.Fields{"urn ns": urn.Namespace, "resourceType": urn.ResourceType},
-		).Error("msg with unknown ResourceType in URN")
+		o.ackEvent(event, errors.Wrap(ErrInvalidEvent, "msg with unknown ResourceType in URN: "+err.Error()))
 
 		return
 	}
 
-	streamEvent := &ptypes.StreamEvent{URN: urn, Event: msg, Data: data}
+	streamEvent := &ptypes.StreamEvent{URN: urn, Event: event, Data: data}
 
 	switch ptypes.EventUrnNamespace(urn.Namespace) {
 	case ptypes.ServerserviceNamespace:
 		o.eventHandler.ServerserviceEvent(otelCtx, streamEvent)
 	case ptypes.ControllerUrnNamespace:
 		o.eventHandler.ControllerEvent(otelCtx, streamEvent)
 	default:
-		if errAck := msg.Ack(); errAck != nil {
-			o.logger.Warn(errAck)
+		o.ackEvent(event, errors.Wrap(ErrInvalidEvent, "msg with unknown URN namespace ignored: "+urn.Namespace))
+	}
+}
+
+func (o *Orchestrator) ackEvent(event events.Message, err error) {
+	if err != nil {
+		// attempt to format the event as JSON
+		// so its easier to read in the logs.
+		logFields := logrus.Fields{"err": err}
+		data, err := json.Marshal(event)
+		if err == nil {
+			logFields["event"] = data
+		} else {
+			logFields["event"] = event
 		}
 
-		o.logger.WithFields(
-			logrus.Fields{"err": err.Error(), "urn ns": urn.Namespace},
-		).Error("msg with unknown URN namespace ignored")
+		// TODO: add metrics for dropped events
+		o.logger.WithError(err).WithFields(logFields).Error("event with error ack'ed")
```

Comment:
do we need the `err: err.Error()` in `Fields` if we call the logger's `WithError` function? That seems redundant.